### PR TITLE
Add support to import predeclared langlibs manually for testable packages

### LINKED ISSUE #25990: Importing Langlib shows different results when running tests individually and within a project
**Description:**
Having `import ballerina/lang.'int;` causes an error `redeclared symbol 'int'` when used in a test file within a module. However this does not happen when the test is run as a .bal file. 

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #25990

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by MaryamZi — COMMENTED** [review]
@chiranSachintha, as discussed offline let's check if we can unify the behaviour of predeclared imports and the user having to specify the imports in both the source and the tests.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 1]
```diff
@@ -415,7 +416,7 @@ public BSymbol resolvePrefixSymbol(SymbolEnv env, Name pkgAlias, Name compUnit)
             entry = entry.next;
         }
 
-        if (env.enclEnv != null) {
+        if (env.enclEnv != null && !(env.node instanceof BLangTestablePackage)) {
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Removed this check with new logic.

File: tests/testerina-integration-test/src/test/resources/project-based-tests/predeclared-import-tests/src/importTests/tests/imports-tests.bal [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/test;
+import ballerina/lang.'float;
+import ballerina/lang.'int;
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply tests/testerina-integration-test/src/test/resources/project-based-tests/predeclared-import-tests/src/importTests/tests/imports-tests.bal:19]: Done.

**Comment by chiranSachintha** [comment]
> @chiranSachintha, as discussed offline let's check if we can unify the behaviour of predeclared imports and the user having to specify the imports in both the source and the tests.

Changed logic to define predeclared imports compilation unit wise.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 3]
```diff
@@ -809,18 +807,23 @@ public void visit(BLangImportPackage importPkgNode) {
...
-            ScopeEntry entry = this.env.scope.lookup(alias);
-            if (entry == NOT_FOUND_ENTRY) {
-                this.env.scope.define(alias, predeclaredModules.get(alias));
-            } else {
-                while (entry.next != NOT_FOUND_ENTRY) {
-                    entry = entry.next;
+        for (BLangCompilationUnit compUnit : compUnits) {
+            for (Name alias : predeclaredModules.keySet()) {
+                BPackageSymbol symbol = duplicatePackagSymbol(predeclaredModules.get(alias));
+                symbol.compUnit = new Name(compUnit.name);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:817]: We don't have to do this if the user has an import with this alias, right? Shall we move these two lines to L820?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 4]
```diff
@@ -809,18 +807,23 @@ public void visit(BLangImportPackage importPkgNode) {
...
+                ScopeEntry entry = this.env.scope.lookup(alias);
+                if (entry == NOT_FOUND_ENTRY) {
+                    this.env.scope.define(alias, symbol);
+                } else {
+                    while (entry.next != NOT_FOUND_ENTRY) {
+                        entry = entry.next;
+                    }
+                    entry.next = new ScopeEntry(symbol, NOT_FOUND_ENTRY);
                 }
-                entry.next = new ScopeEntry(predeclaredModules.get(alias), NOT_FOUND_ENTRY);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:823]: How about changing this to something like

```java
                if (entry == NOT_FOUND_ENTRY) {
                    this.env.scope.define(alias, symbol);
                    continue;
                }

                while (entry.next != NOT_FOUND_ENTRY) {
                    entry = entry.next;
                }
                entry.next = new ScopeEntry(symbol, NOT_FOUND_ENTRY);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 5]
```diff
@@ -809,18 +807,33 @@ public void visit(BLangImportPackage importPkgNode) {
...
         this.env = env;
-        for (Name alias : predeclaredModules.keySet()) {
-            ScopeEntry entry = this.env.scope.lookup(alias);
-            if (entry == NOT_FOUND_ENTRY) {
-                this.env.scope.define(alias, predeclaredModules.get(alias));
-            } else {
-                while (entry.next != NOT_FOUND_ENTRY) {
-                    entry = entry.next;
+        for (BLangCompilationUnit compUnit : compUnits) {
+            for (Name alias : predeclaredModules.keySet()) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:815]: @chiranSachintha, shall we check if we can switch the loops as discussed during the call?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 3 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:817]: Let's try moving this out to a method and calling it only where necessary.

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/ImportTest.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+package org.ballerinalang.testerina.test;
+
+import org.ballerinalang.test.context.BMainInstance;
+import org.ballerinalang.test.context.BallerinaTestException;
+import org.ballerinalang.test.context.LogLeecher;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Test class to test predeclared langlibs using a ballerina project.
```
Comment (MaryamZi) [inline tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/ImportTest.java:28]: Missing the `@since` annotation.

Comment (chiranSachintha) [reply tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/ImportTest.java:29]: Done.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 3 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Done.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 5 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:814]: We can switch two loops but still, we need to check whether alias is defined or not after the first iteration. So do we need to switch these two loops?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:814]: What about doing something like

```java
        for (Name alias : predeclaredModules.keySet()) {
            int index = 0;
            ScopeEntry entry = this.env.scope.lookup(alias);
            if (entry == NOT_FOUND_ENTRY) {
                this.env.scope.define(alias, dupPackageSymbolAndSetCompUnit(predeclaredModules.get(alias),
                            new Name(compUnits.get(index++).name)));
            }
            for (int i = index; i < compUnits.size(); i++) {
                boolean isUndefinedModule = true;
                ...... 
            } 
        }
```

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Changed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 4 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Refactor `initPredeclaredModules` function as mentioned in https://github.com/ballerina-platform/ballerina-lang/pull/26101#discussion_r505188467

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/ImportTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+package org.ballerinalang.testerina.test;
+
+import org.ballerinalang.test.context.BMainInstance;
+import org.ballerinalang.test.context.BallerinaTestException;
+import org.ballerinalang.test.context.LogLeecher;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * @since Swan Lake
```
Comment (MaryamZi) [inline tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/ImportTest.java:28]: Needs to come after the description.

Comment (chiranSachintha) [reply tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/ImportTest.java]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 8]
```diff
@@ -802,25 +800,40 @@ public void visit(BLangImportPackage importPkgNode) {
...
-                this.env.scope.define(alias, predeclaredModules.get(alias));
-            } else {
+            if (entry == NOT_FOUND_ENTRY && !compUnits.isEmpty()) {
+                this.env.scope.define(alias, dupPackageSymbolAndSetCompUnit(predeclaredModules.get(alias),
+                        new Name(compUnits.get(index++).name)));
+                entry = this.env.scope.lookup(alias);
+            }
+            for (int i = index; i < compUnits.size(); i++) {
+                boolean isUndefinedModule = true;
+                if (((BPackageSymbol) entry.symbol).compUnit.value.equals(compUnits.get(i).name)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:823]: Shall we extract `compUnits.get(i).name` out to a variable?

**Review by MaryamZi — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 8 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Fixed.

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**